versioned/storage/cache/src/main/java/org/projectnessie/versioned/storage/cache/CachingPersistImpl.java

```diff
@@ -158,23 +153,17 @@ public boolean[] storeObjs(@jakarta.annotation.Nonnull @Nonnull Obj[] objs)
 
   @Override
   public void upsertObj(@jakarta.annotation.Nonnull @Nonnull Obj obj) throws ObjTooLargeException {
-    try {
-      persist.upsertObj(obj);
-    } finally {
-      cache.remove(obj.id());
-    }
+    persist.upsertObj(obj);
+    cache.put(obj);
```

Comment:
What if `upsertObj` fails? Would we not want to remove the possible old entry from the cache in that case (same for `upsertObjs`)? We cannot be certain what the next `get` will return because we do not know the nature of failure.

Reply:
Similar for `upsertObjs`

Reply:
My initial assumption was that If `upsertObj` fails, then we can be certain that the object has _not_ been persisted. In which case, the most correct thing to do is nothing, i.e. the cache and the underlying store both should keep the previous version of the object, if any, so that they remain in sync.

Thus, I suppose that you are instead considering that a hard failure in `upsertObj` might still have been able to persist the object, in which case, indeed we don't know for sure which version the underlying store has: the previous one (if any), or the current one.

That's a rather pessimistic view, but I am not against it as we never know how future backends could implement `upsertObj`. 

However, if we remove the cache entry on a hard failure for `upsertObj(s)`, I don't see why we wouldn't do the same for `storeObj(s)` as well. Wdyt?

Reply:
`upsertObj` may or may not have done something. But since `upsertObj` is only used during repo-import, it's okay to pessimistically remove it from the cache. `upsertObj` is not meant to be used in production code, because we never update existing objects.

Reply:
Do we benefit from the cache during import _after_ upserts are done?

Reply:
My point is: `upsertObj` is irrelevant for production performance.

> How about ... remove on exception?.. however this will probably make the code more complex 🤔

This is mostly what it's been doing before?

Not sure whether it's really worth to look into a function that's rarely used, if ever used.

Reply:
If upsert is only for imports, shall we refactor the `Persist` interface to make it explicit (maybe have a sub-interface)?

Reply:
> I don't see how upsert's broken ("fix upsert")

I agree that upsert is not broken in current code, while it may be considered not efficient... nonetheless, figuring it all out requires knowledge from code areas far from the cache :) (such as import code).